### ISSUE TITLE
[RFC] Run Loop Helpers

### ISSUE THREAD

**Issue by offirgolan** [machty/ember-concurrency#121]
I think it would be super helpful so include some way to deal with Ember's run loop. There are times where you need to wait for a specific queue in the run loop to execute before you continue with the task. 

## Example 1: 

Adding data incrementally and waiting until `afterRender` to add the next set. 

```js
function * () {
  const model = this.get('model');
  const paginatedModel = paginateModel(model, 10);
  const table = this.get('table')
  
  for (let page of paginatedModel) {
    table.pushRows(page);
    yield waitForQueue('afterRender');
  }
}
```

## Example 2:

Add data to a loading modal only after it has been rendered. 

```js
function * () {
  this.setProperties({ modalLoading: true, showModal: true });
  yield waitForQueue('afterRender');
  this.setProperties({ modalLoading: false, modalData: 'foo bar' });
}
```

## Example 3:

Force the task to wait until after a specific queue has been run. This can be good if a task is called on `init` of a component but requires the dom to be ready.

```js
function * () {
  yield waitForQueue('afterRender');
  this.$().html('Hello World');
}
```

**Comment by machty** [comment]
This is cool :)

I've definitely been interested in having something like this for a while but I always waffled on coming up with an API. I'm glad you've validated that this stuff would be nice to have.

I think `waitForQueue` works. 

I think we should also add an equivalent for awaiting Ember.Evented events. 

```
function * () {
  let event = yield untilEvent('click');
  let payload = yield untilEvent('customEventName');
}
```

So `waitForQueue`/`waitForEvent`, or `untilQueue` or `untilEvent`? I kinda like how `yield untilQueue` reads, but I've found not many people agree w me.

**Comment by offirgolan** [comment]
@machty Im glad you like it! 

I really like the idea of also having the ability to yield of `Ember.Evented` events. That can be super handy!

## waitForQueue

Besides the naming convention, one issue is that the ability to cancel `run.schedule` calls was only introduced in Ember 2.8 (emberjs/ember.js#14550). I'm in the middle of introducing `scheduleTask` to ember-lifeline so I believe the solution to this will be closely related. 

If a workaround can't be created, then I guess we can add an assert in this helper so we only support Ember 2.8+. 

From some random ember-twiddle work, the implementation to get this working is actually fairly simple. 

```js
// Will only work for Ember 2.8+ as of right now
function waitForQueue(q) {
  let timerId;
  let promise = new Ember.RSVP.Promise(r => {
    timerId = Ember.run.schedule(q, r);
  });

  promise.__ec_cancel__ = () => {
    Ember.run.cancel(timerId);
  };
  return promise;
}
```

Usage for it would be something like

```js
yield waitForQueue('afterRender');
```

### waitForEvent

Here is the implementation for `waitForEvent`

```js
function isEvented(c) {
  return c && typeof c.one === 'function' && typeof c.off === 'function';
}

function waitForEvent(c, e) {
  let timerId;
  let fn;
  
  Ember.assert(`${c} must include Ember.Evented to be able to use \`waitForEvent\``, isEvented(c))

  let promise = new Ember.RSVP.Promise(r => {
    fn = r;
    c.one(e, fn);
  });

  promise.__ec_cancel__ = () => {
    c.off(e, fn);
  };
  return promise;
}
```

Usage for it would be something like

```js
yield waitForEvent(this, 'click');
```

## Naming Conventions
I'm kind of open for whatever here. I think the latter 2 roll off the tongue better imho.

```js
yield waitForQueue();
yield waitForEvent();
```

```js
yield untilQueue();
yield unitlEvent();
```

```js
yield forQueue();
yield forEvent();
```
